File: simple-publisher-subscriber/src/listener.cpp
#include "ros/ros.h"            // Include the main ROS header file
#include "std_msgs/String.h"    // Include the String message type from std_msgs

/**
 * This callback funciton is called whenever a new message is recieved on the "chatter" topic.
 * The recieved message is passed as a shared pointer (ConstPtr) to the function. 
 * It logs the message data to the ROS console.
 */
void chatterCallback(const std_msgs::String::ConstPtr& msg)
{
    // Log the recieved message to the ROS console
    ROS_INFO("I heard [%s]", msg->data.c_str());
}

/**
 * The main function initializes the ROS system, subscribes to the "chatter" topic, 
 * and enters a loop, pumping callbacks.
 */
int main(int argc, char **argv)
{
    ros::init(argc, argv, "listener");

    ros::NodeHandle n;

    ros::Subscriber sub = n.subscribe("chatter", 1000, chatterCallback);

    ros::spin();

    return 0;
}

File: simple-publisher-subscriber/scripts/listener.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import String

def callback(data):
    """Receives messages from the 'chatter' topic and logs them.
    
    Args:
        data (std_msgs.msg.String): The message received from the 'chatter' topic.
    """
    rospy.loginfo(rospy.get_caller_id() + " I heard %s", data.data)

def listener():
    """Initializes a ROS node named 'listener' and subscribes to the 'chatter' topic.
    """
    # Initialize the ROS node named 'listener'
    rospy.init_node('listener', anonymous=True)

    # Create a subscriber that will recieve messages from the 'chatter' topic
    rospy.Subscriber('chatter', String, callback)

    # Simply keeps python from exiting until this node is stopped
    rospy.spin()

if __name__ == '__main__':
    try:
        listener()
    except rospy.ROSInterruptException:
        pass